Add basic prove_refinement tactic

m-yac opened the pull request:
This PR adds a `prove_refinement` tactic which can be used to prove goals of the form:
```
(a1:A1) -> ... (an:An) -> refinesS_eq ...
```
within a call to `prove_extcore` (the `Prop` version of `prove_print`). All of the existing MRSolver unit tests work with this new tactic, and `prove_refinement` versions of all tests in `mr_solver_unit_tests.saw` have been added to that file to reflect this.

Note this tactic hooks into the existing MRSolver machinery with no changes, so if MRSolver fails it will currently exit the SAW script instead of proving more useful feedback and/or leaving unsolved goals over for the user. There is also currently no way to use theorems proved using this tactic in subsequent proofs using this tactic. Both of these things will change in coming PRs as we adapt MRSolver to this new paradigm.

RyanGlScott commented on src/SAWScript/Interpreter.hs:
```diff
@@ -3780,6 +3785,12 @@ primitives = Map.fromList
     , " 1 = basic debug output, 2 = verbose debug output,"
     , " 3 = all debug output" ]
 
+  , prim "prove_refinement" "ProofScript ()"
```
Naming quibble: most (and quite possibly all) of the existing `prove_*` functions in SAWScript inhabit `TopLevel`, as they meant to be the things that actually perform proofs. `prove_refinement` is an exception to this convention, as it is a proof _script_, not a thing that performs the proof on its own.

How would you feel about simply calling this "`refinement`"?

m-yac replied:
Ah, that's an excellent point!

I looked at a few other `ProofScript` commands in `Interpreter.hs`, and it looks like the pattern is that the Haskell functions `proveZ3`, `proveRME`, etc. map to the SAW commands `z3`, `rme`, etc. – so following that, indeed our `proveRefinement` should indeed be `refinement` in SAW.

However, a command named `refinement` really doesn't strike me as the right thing here. Maybe `mrsolver`? `solve_refinement`? @eddywestbrook Do you have thoughts? (Regardless, I agree we shouldn't start our command with `prove_`.)

RyanGlScott replied:
We also have a variety of other `ProofScript` commands that begin with `goal_*` (e.g., `goal_eval`), so `goal_refinement` is another option.

eddywestbrook replied:
After some discussion, Matt and I decided on `mrsolver` as the command name. That seems to fit with how tactics for calling other solvers and provers are named, as the name of the solver itself.

RyanGlScott commented on src/SAWScript/Proof.hs:
```diff
@@ -1072,6 +1076,9 @@ data Evidence
     --   sequent calculus axiom, which connects a hypothesis to a conclusion.
   | AxiomEvidence
 
+  -- | TODO: Add some evidence
```
This TODO aside, it would be worth including some documentation about how `MrSolverEvidence` can be produced.

m-yac commented on saw-core/prelude/Prelude.sawcore:
```diff
@@ -3294,6 +3307,74 @@ appendCastBVVecS E stack n len1 len2 len3 a v1 v2 =
+--------------------------------------------------------------------------------
+-- Monadic test functions
+
+test_fun0 : Vec 64 Bool -> SpecM VoidEv emptyFunStack (Vec 64 Bool);
```
No, you're right, these really should not be in the Prelude. My only guess is maybe the were put there before we had the `load_sawcore_from_file` command? Regardless – I've moved them to a their own `.sawcore` file which gets imported via `load_sawcore_from_file` in `mr_solver_unit_tests.saw`.

m-yac commented:
Hunh, and now it looks like the CI refuses to run for some reason on this latest commit (that checkmark is bogus and appeared instantly). We should definitely make it sure it runs before merging, I had a GHC 9 issue on my previous commit that I should have fixed, but didn't actually check on my machine.

eddywestbrook commented on saw-core/prelude/Prelude.sawcore:
```diff
@@ -2989,10 +2928,88 @@ emptyFunStack = Nil1 (List1 LetRecType);
 pushFunStack : List1 LetRecType -> FunStack -> FunStack;
 pushFunStack frame stack = Cons1 (List1 LetRecType) frame stack;
 
+-- The type of FunStackE
```
Maybe add a little more documentation about what a `FunStackE` is, by describing that it's either an error, an `E`, or a `FrameCall`

eddywestbrook commented on saw-core/prelude/Prelude.sawcore:
```diff
@@ -2989,10 +2928,88 @@ emptyFunStack = Nil1 (List1 LetRecType);
 primitive SpecM : (E:EvType) -> FunStack -> sort 0 -> sort 0;
 
 
+-- The type of the preconditions needed for refinesS
```
Explain in the comments that this is the type of a relation between `FunStackE` objects

eddywestbrook commented on saw-core/prelude/Prelude.sawcore:
```diff
@@ -2989,10 +2928,88 @@ emptyFunStack = Nil1 (List1 LetRecType);
+SpecPreRel E1 E2 stack1 stack2 =
+  FunStackE_type E1 stack1 -> FunStackE_type E2 stack2 -> Prop;
+
+-- The type of the postconditions needed for refinesS
```
Similarly to the above, explain what this is

eddywestbrook reviewed (approved):
Looks good, other than the few places I've asked for more comments in the SAW core. I'm marking as approved because those are minor, but please address those before merging.

m-yac commented:
Okay, @eddywestbrook, I think I addressed your comments! Thank you for suggesting to add more docs, those definitions are definitely not intuitive at first glance.

Also thankfully the CI now appears to start when I commit now, not sure what that one failure was about ¯\_(ツ)_/¯

eddywestbrook commented:
Wuh oh, looks like the Coq translator doesn't want to translate the `List1` recursor. The hack we use for all the other recursors is to define `List1__rec` and add a special case for it to the Coq special treatment. But please file a bug about this, if there isn't one already, because that shouldn't be necessary.

m-yac commented:
No idea why `mr_solver_unit_tests.saw` suddenly failed on my last commit – the only changes I made were to `saw-core-coq`, which shouldn't affect that test, and on the previous commit that test passed. I also couldn't replicate the failure on my machine. So, I'm trying out pushing a merge commit and hoping that a new CI run somehow magically fixes it.

My only guess is that somehow it's running the test with an old version of `saw` that doesn't have `test_fun0` removed from the Prelude. That would explain why the error is that `test_fun0` is being added twice, but I have no idea why the CI wouldn't use the version of `saw` it built on this branch...
